apps/emqx_rule_engine/src/emqx_rule_registry.erl

```diff
@@ -471,11 +477,16 @@ code_change(_OldVsn, State, _Extra) ->
 
 get_all_records(Tab) ->
     %mnesia:dirty_match_object(Tab, mnesia:table_info(Tab, wild_pattern)).
-    ets:tab2list(Tab).
+    {atomic, Ret} =
+        ekka_mnesia:ro_transaction(?RULE_ENGINE_SHARD,
+                                   fun() ->
+                                           ets:tab2list(Tab)
```

Comment:
A part from running on core node or replicant, I think ets:tab2list/1 won't play well in transcational context or? Isn't it is truely a dirty context?

Reply:
True. I wrapped this in r/o transaction (that runs locally on the replicant) to avoid reading data from the table while the shard is being bootstrapped. The contents of the table during bootstrap are inconsistent.
In this case R/O transaction will restart when it detects this situation.

Reply:
P.S. I will add a comment